org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/text/correction/SourceAssistProcessor.java

```diff
@@ -618,6 +638,64 @@ private Optional<Either<Command, CodeAction>> getFinalModifierWherePossibleActio
 		}
 	}
 
+	private Optional<Either<Command, CodeAction>> getSortMembersProposal(IInvocationContext context, CodeActionParams params, String kind, String label) {
+		CategorizedTextEditGroup group = new CategorizedTextEditGroup(label, new GroupCategorySet(new GroupCategory(label, label, label)));
+		try {
+			TextEdit edit = CompilationUnitSorter.sort(context.getASTRoot(), new DefaultJavaElementComparator(false), 0, group, null);
+			if (edit == null) {
+				return Optional.empty();
+			}
+			CodeActionProposal sortMembersProposal = (pm) -> {
+				return convertToWorkspaceEdit(context.getCompilationUnit(), edit);
+			};
+			return getCodeActionFromProposal(params.getContext(), context.getCompilationUnit(), label, kind, sortMembersProposal, CodeActionComparator.SORT_MEMBERS_PRIORITY);
+		} catch (JavaModelException e) {
+			return Optional.empty();
+		}
+	}
+
+	private Optional<Either<Command, CodeAction>> getSortMembersForSelectionProposal(IInvocationContext context, CodeActionParams params, List<ASTNode> coveredNodes) {
+		String label = ActionMessages.SortMembers_selectionLabel;
+		CategorizedTextEditGroup group = new CategorizedTextEditGroup(label, new GroupCategorySet(new GroupCategory(label, label, label)));
+		PartialSortMembersOperation operation = new PartialSortMembersOperation(new IJavaElement[] { context.getASTRoot().getJavaElement() }, new DefaultJavaElementComparator(false));
+		try {
+			TextEdit edit = operation.calculateEdit(context.getASTRoot(), coveredNodes, group);
+			if (edit == null) {
+				return Optional.empty();
+			}
+			CodeActionProposal sortMembersProposal = (pm) -> {
+				return convertToWorkspaceEdit(context.getCompilationUnit(), edit);
+			};
+			return getCodeActionFromProposal(params.getContext(), context.getCompilationUnit(), label, JavaCodeActionKind.QUICK_ASSIST, sortMembersProposal, CodeActionComparator.SORT_MEMBERS_PRIORITY);
+		} catch (JavaModelException e) {
+			return Optional.empty();
+		}
+	}
+
+	private Optional<Either<Command, CodeAction>> getSortMembersAction(IInvocationContext context, CodeActionParams params) {
+		CompilationUnit unit = context.getASTRoot();
+		if (unit != null) {
+			List<AbstractTypeDeclaration> types = unit.types();
+			if (types.size() == 1) {
+				String typeName = CodeActionUtility.getTypeName(types.get(0));
+				if (typeName != null) {
+					return getSortMembersProposal(context, params, JavaCodeActionKind.SOURCE_SORT_MEMBERS, Messages.format(ActionMessages.SortMembers_templateLabel, typeName));
+				}
+			} else if (types.size() > 1) {
+				return getSortMembersProposal(context, params, JavaCodeActionKind.SOURCE_SORT_MEMBERS, ActionMessages.SortMembers_label);
+			}

```

Comment:
In places where you need to get the name of the compilation unit's type, does `CompilationUnit#getTypeRoot().getElementName()` satisfy your needs ?

For figuring out if the compilation unit declares more than 1 type, for the purpose of using the correct label, could you use `context.getCompilationUnit().getAllTypes().length > 1` ?

Reply:
You're right, I just made a mistake on this, let me correct it in the next commit.

Reply:
Since we will always sort the whole Java file regardless of the length of types (eclipse does the similar thing), I think we can just use `unit.getTypeRoot().getElementName()` for formatting the label here.